cdap-api/src/main/java/co/cask/cdap/internal/workflow/DefaultWorkflowActionSpecification.java

```diff
@@ -35,30 +39,43 @@
   private final String name;
   private final String description;
   private final Map<String, String> properties;
+  private final Set<String> datasets;
 
-  public DefaultWorkflowActionSpecification(String name, String description, Map<String, String> properties) {
-    this(null, name, description, properties);
+  public DefaultWorkflowActionSpecification(String name, String description,
+                                            Map<String, String> properties, Set<String> datasets) {
+    this.className = null;
+    this.name = name;
+    this.description = description;
+    this.properties = properties;
+    this.datasets = ImmutableSet.copyOf(datasets);
   }
 
   public DefaultWorkflowActionSpecification(WorkflowAction action) {
     WorkflowActionSpecification spec = action.configure();
 
     Map<String, String> properties = Maps.newHashMap(spec.getProperties());
+    Set<String> datasets = Sets.newHashSet();
     Reflections.visit(action, TypeToken.of(action.getClass()),
+                      new DataSetFieldExtractor(datasets),
                       new PropertyFieldExtractor(properties));
 
+    // Add datasets that are specified in overriding configure with useDataset method
+    datasets.addAll(spec.getDatasets());
```

Comment:
shouldn't the `datasets` be populated before the `Reflections.visit` ? - 
`Set<String> datasets = Sets.newHashSet(spec.getDatasets());`